### PR TITLE
Feat: Make sure 'sqlmesh run' evaluates against a finalized environment

### PR DESCRIPTION
**Pull request by izeigerman**

### CODE REVIEW

File: sqlmesh/core/context.py [hunk 1]
```diff
@@ -450,22 +452,80 @@ def run(
...
+            plan_id_at_start = _block_until_finalized()
+
+            def _has_environment_changed() -> bool:
+                assert environment is not None
+                current_environment_state = self.state_sync.get_environment(environment)
+                return (
+                    not current_environment_state
+                    or current_environment_state.plan_id != plan_id_at_start
+                    or not current_environment_state.finalized_ts
+                )
```
Comment (eakmanrq) [inline sqlmesh/core/context.py:499]: Is this used?

File: sqlmesh/core/scheduler.py [hunk 2]
```diff
@@ -193,6 +193,7 @@ def run(
         restatements: t.Optional[t.Dict[str, SnapshotInterval]] = None,
         ignore_cron: bool = False,
         selected_snapshots: t.Optional[t.Set[str]] = None,
+        circuit_breaker: t.Optional[t.Callable[[], bool]] = None,
```
Comment (eakmanrq) [inline sqlmesh/core/scheduler.py:196]: Is this used?

File: sqlmesh/core/context.py [hunk 3]
```diff
@@ -450,22 +452,80 @@ def run(
...
+
+        env_check_attempts_num = max(
+            1,
+            self.config.run.environment_check_max_wait
+            // self.config.run.environment_check_interval,
+        )
+
+        def _block_until_finalized() -> str:
+            for _ in range(env_check_attempts_num):
+                assert environment is not None
```
Comment (tobymao) [inline sqlmesh/core/context.py:471]: is this for mypy?

File: sqlmesh/core/context.py [hunk 4]
```diff
@@ -450,22 +452,80 @@ def run(
...
+                    return environment_state.plan_id
+                logger.warning(
+                    "Environment '%s' is being updated by plan '%s'. Retrying in %s seconds...",
+                    environment,
+                    environment_state.plan_id,
+                    self.config.run.environment_check_interval,
+                )
+                time.sleep(self.config.run.environment_check_interval)
+            raise SQLMeshError(
+                f"Exceeded the maximum wait time for environment '{environment}' to become ready."
```
Comment (tobymao) [inline sqlmesh/core/context.py:485]: link to the documentation on how to change these settings and what it means when this error is hit

File: sqlmesh/core/context.py (hunk 1 above)
Comment (izeigerman) [reply sqlmesh/core/context.py:499]: Thanks, good catch. Still testing this

File: sqlmesh/utils/errors.py [hunk 5]
```diff
@@ -119,6 +119,11 @@ class UnsupportedCatalogOperationError(EngineAdapterError):
     pass
 
 
+class CircuitBreakerError(SQLMeshError):
+    def __init__(self) -> None:
+        super().__init__("Circuit breaker has been triggered.")
```
Comment (tobymao) [inline sqlmesh/utils/errors.py:124]: this error is a bit cryptic

**Review by tobymao — COMMENTED** [review]
z

File: sqlmesh/core/context.py (hunk 3 above)
Comment (izeigerman) [reply sqlmesh/core/context.py]: correct

File: sqlmesh/core/scheduler.py (hunk 2 above)
Comment (izeigerman) [reply sqlmesh/core/scheduler.py:196]: yes, we now pass `_has_environment_changed` as a circuit breaker

File: sqlmesh/utils/errors.py (hunk 5 above)
Comment (izeigerman) [reply sqlmesh/utils/errors.py:124]: It should never be user-facing though. The idea is that it's an internal mechanics and someone's always expected to intercept it and provide a friendlier explanation.

File: sqlmesh/core/context.py (hunk 4 above)
Comment (izeigerman) [reply sqlmesh/core/context.py]: improved the error message

**Review by eakmanrq — APPROVED** [review]
👏

**Pull request merged**